airbyte-api/src/main/openapi/config.yaml

```diff
@@ -3980,15 +4015,134 @@ components:
-        state:
-          $ref: "#/components/schemas/ConnectionStateObject"
-    ConnectionStateObject:
+        state: # legacy state object
+          $ref: "#/components/schemas/StateBlob"
+        streamState:
+          type: array
+          items:
+            $ref: "#/components/schemas/StreamState"
+        globalState:
+          $ref: "#/components/schemas/AirbyteGlobalState"
+    StateBlob:
+      type: object
+    StreamState:
+      type: object
+      required:
+        - streamDescriptor
+      properties:
+        streamDescriptor:
+          $ref: "#/components/schemas/StreamDescriptor"
+        streamState:
+          $ref: "#/components/schemas/StateBlob"
+    AirbyteGlobalState:
       type: object
+      required:
+        - streamStates
+      properties:
+        shared_state:
+          $ref: "#/components/schemas/StateBlob"
+        streamStates:
+          type: array
+          items:
+            $ref: "#/components/schemas/StreamState"
+    ConnectionStateType:
+      type: string
+      enum:
+        - global
+        - stream
+        - legacy
+        - not_set
+    CatalogDiff:
+      type: object
+      description: "Describes the difference between two Airbyte catalogs."
+      required:
+        - transformations
+      properties:
+        transformations:
+          description: list of stream transformations. order does not matter.
+          type: array
+          items:
+            $ref: "#/components/schemas/StreamTransform"
+    StreamTransform:
+      type: object
+      required:
+        - transformType
+      properties:
+        transformType:
+          type: string
+          enum:
+            - add_stream
+            - remove_stream
+            - update_stream
+        addStream:
+          $ref: "#/components/schemas/StreamDescriptor"
+        removeStream:
+          $ref: "#/components/schemas/StreamDescriptor"
+        updateStream:
+          type: array
+          description: list of field transformations. order does not matter.
+          items:
+            $ref: "#/components/schemas/FieldTransform"
+    FieldTransform:
+      type: object
+      description: "Describes the difference between two Streams."
+      required:
+        - transformType
+      properties:
+        transformType:
+          type: string
+          enum:
+            - add_field
+            - remove_field
+            - update_field_type
+        addField:
+          $ref: "#/components/schemas/FieldNameAndType"
+        removeField:
+          $ref: "#/components/schemas/FieldNameAndType"
+        updateFieldType:
+          $ref: "#/components/schemas/FieldTypeUpdate"
+    FieldNameAndType:
+      type: object
+      required:
+        - fieldName
+        - fieldType
+      properties:
+        fieldName:
+          type: array
+          items:
+            type: string
+        fieldType:
+          type: array
+          items:
+            type: string
```

Comment:
From the discussion I had with Tim, this should be a schema and be an object with no constraint (i.e. `additionalProperties = true`).

Reply:
If this is being changed to a schema, maybe we should also rename these properties from `type` -> `schema`, e.g. `update_field_schema`, `fieldSchema`, `FieldSchemaUpdate`, etc.

Reply:
agreed.